callable/src/main/java/io/grpc/callable/Callable.java

```diff
@@ -0,0 +1,423 @@
+   * Callable.Transformer&lt;Book, GetAuthorRequest> bookToGetAuthorRequest = ...;
+   * Author response =
+   *     Callable.create(LibraryGrpc.CONFIG.getBook)
+   *             .followedBy(Callable.create(bookToGetAuthorRequest))
+   *             .followedBy(Callable.create(LibraryGrpc.CONFIG.getAuthor))
+   *             .call(channel, new GetBookRequest().setName(bookName).build());
+   * </pre>
+   *
+   * <p>For streaming calls, each output of the first callable will be forwarded to the second
+   * one as it arrives, allowing for streaming pipelines.
+   */
+  public <FinalResT> Callable<RequestT, FinalResT>
+      followedBy(Callable<ResponseT, FinalResT> callable) {
+    return new FollowedByCallable<RequestT, ResponseT, FinalResT>(this, callable);
+  }
+
+  // Channel Redirection
+  // ===================
+
+  /**
+   * Returns a callable which redirects the execution of this callable to a channel determined at
+   * call execution time, via a supplier. This can be used for creating batches which target
+   * different services. Example:
+   *
+   * <pre>
+   * Author response =
+   *     Callable.redirect(
+   *       Suppliers.ofInstance(bookChannel),
+   *       Callable.create(LibraryGrpc.CONFIG.getBook))
+   *     .followedBy(Callable.create(LibraryGrpc.CONFIG.getAuthor))
+   *     .call(authorChannel, new GetBookRequest().setName(bookName));
+   * </pre>
+   */
+  public static <RequestT, ResponseT> Callable<RequestT, ResponseT>
+      redirect(final Supplier<Channel> channelSupplier,
+               final Callable<RequestT, ResponseT> scoped) {
+    return new Callable<RequestT, ResponseT>() {
+      @Override
+      public ClientCall<RequestT, ResponseT> newCall(Channel channel) {
+        return scoped.newCall(channelSupplier.get());
+      }
+    };
+  }
+
+  // Retrying
+  // ========
+
+  /**
+   * Returns a callable which retries using exponential back-off on transient errors. Example:
+   *
+   * <pre>
+   * Response response = Callable.create(METHOD).retrying().call(channel, request);
+   * </pre>
+   *
+   * <p>The call will be retried if and only if the returned status code is {@code UNAVAILABLE}.
+   *
+   * <p>No output will be produced until the underlying callable has succeeded. Applied to compound
+   * callables, this can be used to implement simple transactions supposed the underlying callables
+   * are either side-effect free or idempotent.
+   *
+   * <p>Note that the retry callable requires to buffer all inputs and outputs of the underlying
+   * callable, and should be used with care when applied to streaming calls.
+   */
+  public Callable<RequestT, ResponseT> retrying() {
+    return new RetryingCallable<RequestT, ResponseT>(this);
+  }
+
+  // Page Streaming
+  // ==============
+
+  /**
+   * Returns a callable which streams the resources obtained from a series of calls to a method
+   * implementing the pagination pattern. Example:
+   *
+   * <pre>
+   *  for (Resource resource :
```

Comment:
Do we really want this over a 'forEach' pattern

https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#forEach-java.util.function.Consumer-

It doesn't seem any more succinct so I'm not sure it's valuable as a first class part of the API or do we consider these utility functions and we expect to provide a bunch of them

We can also support forEachWithExector in case the processing doesn't have to happen in sequence